lib/services/local/trust.go

```diff
@@ -160,48 +218,74 @@ func (s *CA) CompareAndSwapCertAuthority(new, expected types.CertAuthority) erro
-	// types.CertAuthority as well if they exist.
-	err := s.Delete(ctx, backend.Key(authoritiesPrefix, deactivatedPrefix, string(id.Type), id.DomainName))
-	if err != nil {
-		if !trace.IsNotFound(err) {
+	return s.DeleteCertAuthorities(ctx, id)
+}
+
+// DeleteCertAuthorities deletes multiple cert authorities atomically.
+func (s *CA) DeleteCertAuthorities(ctx context.Context, ids ...types.CertAuthID) error {
+	var condacts []backend.ConditionalAction
+	for _, id := range ids {
+		if err := id.Check(); err != nil {
 			return trace.Wrap(err)
 		}
+		for _, key := range [][]byte{activeKey(id), inactiveKey(id)} {
+			condacts = append(condacts, backend.ConditionalAction{
+				Key:       key,
+				Condition: backend.Whatever(),
+				Action:    backend.Delete(),
+			})
+		}
 	}
-	err = s.Delete(ctx, backend.Key(authoritiesPrefix, string(id.Type), id.DomainName))
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	return nil
+
+	_, err := s.AtomicWrite(ctx, condacts)
+	return trace.Wrap(err)
 }
 
 // ActivateCertAuthority moves a CertAuthority from the deactivated list to
 // the normal list.
 func (s *CA) ActivateCertAuthority(id types.CertAuthID) error {
-	item, err := s.Get(context.TODO(), backend.Key(authoritiesPrefix, deactivatedPrefix, string(id.Type), id.DomainName))
-	if err != nil {
-		if trace.IsNotFound(err) {
-			return trace.BadParameter("can not activate cert authority %q which has not been deactivated", id.DomainName)
+	return s.ActivateCertAuthorities(context.TODO(), id)
+}
+
+// ActivateCertAuthorities activates multiple cert authorities atomically.
+func (s *CA) ActivateCertAuthorities(ctx context.Context, ids ...types.CertAuthID) error {
+	var condacts []backend.ConditionalAction
+	var domainNames []string
+	for _, id := range ids {
+		if err := id.Check(); err != nil {
+			return trace.Wrap(err)
 		}
-		return trace.Wrap(err)
-	}
 
-	certAuthority, err := services.UnmarshalCertAuthority(
-		item.Value, services.WithResourceID(item.ID), services.WithExpires(item.Expires), services.WithRevision(item.Revision))
-	if err != nil {
-		return trace.Wrap(err)
-	}
+		if !slices.Contains(domainNames, id.DomainName) {
+			domainNames = append(domainNames, id.DomainName)
+		}
 
-	err = s.UpsertCertAuthority(context.TODO(), certAuthority)
-	if err != nil {
-		return trace.Wrap(err)
-	}
+		item, err := s.Get(ctx, inactiveKey(id))
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return trace.Errorf("can not activate cert authority %q of type %q (not a currently inactive ca)", id.DomainName, id.Type)
+			}
+			return trace.Wrap(err)
+		}
 
-	err = s.Delete(context.TODO(), backend.Key(authoritiesPrefix, deactivatedPrefix, string(id.Type), id.DomainName))
-	if err != nil {
+		condacts = append(condacts, []backend.ConditionalAction{
+			{
+				Key:       inactiveKey(id),
+				Condition: backend.Revision(item.Revision),
+				Action:    backend.Delete(),
+			},
+			{
+				Key:       activeKey(id),
+				Condition: backend.Whatever(),
```

Comment:
```suggestion
				Condition: backend.NotExists(),
```
right?

Reply:
Reasonable arguments for either I think.  I lean toward keeping activation/deactivation as overwriting because otherwise we'd get stuck if we encountered dangling state.  Asserting the revision value on the existing CA ensures that this will conflict with concurrent activation/deactivation attempts, so we're already protected from that.  But if a previous (non-atomic) operation failed in the past s.t. we were left with dangling state, I think overwriting it is preferable to locking up.

Reply:
I think the currently active CA should take priority, so maybe we should deliberately make the check `NotExists()` in `ActivateCertAuthorities` and `Whatever()` in `DeactivateCertAuthorities`?

Reply:
Active CA taking priority sounds reasonable to me.  I'll do that.